### ISSUE TITLE
Is this intended to get merged back?

### ISSUE THREAD

**Issue by drewcovi** [Akylas/nativescript-bluetooth#5]
It seems we have a few mentions of nativescript-akylas-bluetooth, but not in the readme. 

Is there an intent to eventually open a PR with Eddy?

**Comment by rynop** [comment]
Martin - my 2cents would be to PR your new function and re-working back to https://github.com/EddyVerbruggen/nativescript-bluetooth .  It will allow the community to put all of our effort into one BLE lib vs 2 (or more).

You can see my discussion with Eddy referenced above...

**Comment by farfromrefug** [comment]
@drewcovi  @rynop yes i want to merge and i already discussed that with @EddyVerbruggen once.
However it is not that easy as it not only code "breaking" changes but also project structure( and i am not finished). And making PRs takes a lot of time as they need to be clean for @EddyVerbruggen to review and approve them.
Also it would break code for a lot of user as i decided to take a different approach from @EddyVerbruggen on some parts, especially the "send" data parameter format.
The more i think about it the more i think i will go for a new module ```nativescript-bluetooth``` to get a good ratio between time consumption and "sharing" of my fork.

Still if @EddyVerbruggen decide to merge test (which is a lot faster than PRs as he only need to merge / test) i will help maintaining the original module.

**Comment by rynop** [comment]
That is awesome Martin you are open to combining all of our efforts into one plugin.

I totally understand the time it takes to make a clean PR especially when philosophies and project structures have  changed.

With that in mind, I'm a bit confused what you mean when you say:

> i think i will go for a new module `nativescript-bluetooth`

To help me clarify what you are saying here: are you proposing a brand new plugin at a new git repo?

**If yes**: the `nativescript-bluetooth` name is [already taken](https://market.nativescript.org/plugins/nativescript-bluetooth).  In addition, isn't this the scenario we are already in (2 plugin codebases)?  While your (this) repo is forked from `EddyVerbruggen/nativescript-bluetooth`, in my opinion its has diverged enough to really be considered a different plugin.  I think you agree because you registered [nativescript-akylas-bluetooth](https://market.nativescript.org/plugins/nativescript-akylas-bluetooth) as a separate plugin.

**If no**: can you elaborate a bit more on what you meant?

My vote would be for @EddyVerbruggen to make a new `v2` branch on `EddyVerbruggen/nativescript-bluetooth` that is totally empty.  You then fork on that branch, prime it with the current contents of `Akylas/nativescript-bluetooth` and PR.  This is very little work for you (I could even do the work if you like).

We can then all discuss / work on / cleanup / test / document in `EddyVerbruggen/nativescript-bluetooth#v2` branch.  Once its ready to be released, `v2` is merged into `master` and released under semver v2.0.0.  `npm`/`yarn`/`tns plugin` semver enforcement in the `package.json` will make it so no one in the field breaks.  By default `"nativescript-[akylas-]bluetooth": "^1.x.x"` is put into the `package.json` so anyone that updates [won't get](https://michaelsoolee.com/npm-package-tilde-caret/) version `2.x.x`.

@farfromrefug let me know answer to question above and let me know if my proposal sounds OK.  If yes, let me know if you'd like me to do the work of priming `EddyVerbruggen/nativescript-bluetooth#v2` branch

@EddyVerbruggen are you OK with the `EddyVerbruggen/nativescript-bluetooth#v2` branch proposal?  If yes, and Martin is cool with it go ahead and create the `v2` branch.  Martin and I will work out getting the PR to `v2`.

**Comment by EddyVerbruggen** [comment]
Hey guys, I'd love to see our efforts merged and it would be awesome to maintain 1 plugin together. I can add you as co-owners of both the GitHub repo and the npm package.

A 2.0.0 version is inevitable I think. It would be nice to keep backward compatibility as much as possible, but don't worry too much if we need to break a thing or two.

@rynop I've just added you to the GitHub repo as a collaborator. If/when @farfromrefug agrees please don't waste time waiting for me to respond. I'm not always able to reply swiftly. Thanks!

**Comment by farfromrefug** [comment]
@rynop  typo error when I said new plugin I meant nativescript-bt or something. Now I would prefer to work with @EddyVerbruggen. More maintainers is always better and some have more specialties than others (me no experience with beacons)
@EddyVerbruggen of course I want to be a contributor! one question though to make sure we start things right, would you be okay for a new branch with my code as it is ? then we decide together the merge(without name change). that way until 2.0 I can still release my fork (that I use in multiple projects) while commiting  on the final repo.

**Comment by rynop** [comment]
Thanks @EddyVerbruggen, I made [v2](https://github.com/EddyVerbruggen/nativescript-bluetooth/tree/v2).  If you are cool with using nativescript-akylas-bluetooth#master as base for `v2`, then @farfromrefug if you'd please merge https://github.com/Akylas/nativescript-bluetooth/pull/9 I can do the work to prime `v2`.  It will only take me a few seconds.

**Comment by EddyVerbruggen** [comment]
Yep, @farfromrefug, happy to use your code as a starting point for the new branch. I've added you as a collaborator as well. THANKS guys!

**Comment by rynop** [comment]
@farfromrefug no need to merge https://github.com/Akylas/nativescript-bluetooth/pull/9 . I just merged from your upstream and primed https://github.com/EddyVerbruggen/nativescript-bluetooth/tree/v2 with it!

Eddy and Martin I really appreciate your willingness, openness and quickness in making this happen.  No egos, no drama.  I'm confident that everyone will benefit from the community getting behind one NS BLE plugin.

In the coming days I'm going to be implementing a prototype using this plugin.  I'll update docs and examples as I go along.  I'll do so right against `v2` without a PR if that is OK with you guys.  More efficient.  If I have any functional changes/proposals I'll obviously branch and PR.

THANK YOU!

**Comment by drewcovi** [comment]
Thanks a million @rynop @EddyVerbruggen @farfromrefug. Building out a prototype / platform here at my dayjob and this is a huge help. I'm a huge fan of Nativescript and it's promise with web/native code sharing.

**Comment by farfromrefug** [comment]
@rynop  what is it you call a prototype? anyway thanks for your help too!

**Comment by drewcovi** [comment]
On my end I’ve built a POC with a raspberry pi zero and bleno to establish services characteristics and descriptors, and pass WiFi credentials.

After a feature complete demo in Cordova and Ember, I’m migrating to Angular and Rxjs since Nativescript is so much more performance. So thanks @rynop, perusing your demo really helps someone like me dive deeper into rxjs.

We’re an IoT company, with a fairly huge footprint in the states, so I’m super optimistic about how this turns out.

**Comment by farfromrefug** [comment]
@drewcovi  great! on my side I am not really satisfied with bluetooth performances in Nativescript. the fact that everything runs on main thread makes for big performance drawback when using real time communication with devices. I am thinking about writing it natively with a service on android. already have all the codebase as I did it in another framework. that way only connect/disconnect/write.../read events , would be called on main thread. the whole logic would run on a background thread

**Comment by rynop** [comment]
@farfromrefug I've started a company, so I'm in the process of making a native app to prototype some functionality.

@drewcovi you are welcome.  Hopefully everyone in community will benefit.  It's great to have a large company with some resources devoted to POC with NativeScript and this plugin.  Will only make both better.

@farfromrefug I haven't looked at the nativescript-bluetooth code much.  I was under the impression that the majority of the BT code ran natively outside of the JS main event loop thread.  Thats a bummer.  Looking forward to your advancements in this area as what I'm going to be using it for will need as close to real-time comm as possible.

**Comment by farfromrefug** [comment]
@rynop all JS codes in NS are run in main thread. So all delegates methods, bluetooth events ... Which does not really make sense. It's even harder to wrap your head around this when you make it run in a service that you initiate from JS which is what i do. This means that the service running in the background actually kind of ```run``` in the main thread. I am seeing some hick ups when my app is in the background on some devices, and i am guessing it's due to the threading issue.

**Comment by rynop** [comment]
Thanks.

The angular demo in `v2` branch has a self-contained service (disclosure I did not write it): https://github.com/EddyVerbruggen/nativescript-bluetooth/blob/v2/demo-ng/app/services/bluetooth.service.ts 

It leverages angular `NgZone` which might help what you may be seeing.

My prototype will have to solve some of the same problems (I need to `read()` and continuously update the view in as close to real-time as possible,  so I'll make sure to contribute back to the `ng-demo` any best practices/tricks I find).

**Comment by farfromrefug** [comment]
@rynop be careful angular service has nothing to do with Android service!
Angular service wont make your app stay awake when running on the background.
Android services are here to work in the background or even without a running interface
I have a running and working app with it. might try to write a sample app at one point.

**Comment by rynop** [comment]
Ah sorry misinterpreted what you meant by service.  I know on iOS background BLE requires special permissions and [tweaks](https://wojciechkulik.pl/xamarin-ios/bluetooth-low-energy-background-mode-on-ios) in `info.plist`.

A plugin equivalent to this one for react native ([react-native-ble-plx](https://github.com/Polidea/react-native-ble-plx)), supports [background BLE](https://github.com/Polidea/react-native-ble-plx#configuration--installation).  Would be nice if `nativescript-bluetooth` got support for it as well.  Unfortunately I've never done native dev on either platform, so I'm not in a good position to add support.
